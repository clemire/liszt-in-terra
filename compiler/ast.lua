--[[ Module defines all of the AST nodes used to represent the Liszt 
     language.
]]--
local A = {}
package.loaded["compiler.ast"] = A

---------------------------
--[[ Declare AST types ]]--
---------------------------
local AST             = { kind = 'ast', is_liszt_ast = true }
AST.__index           = AST

local LisztKernel     = { kind = 'kernel' }
local Block           = { kind = 'block'  } -- Statement*
  -- store condition and block to be executed for if/elseif clauses
local CondBlock       = { kind = 'condblock' } 

-- Expressions:
local Expression      = { kind = 'expr'   } -- abstract
local BinaryOp        = { kind = 'binop'  }
local Reduce          = { kind = 'reduce' }
local UnaryOp         = { kind = 'unop'   }
local Tuple           = { kind = 'tuple'  }

local TableLookup     = { kind = 'lookup' }
local VectorIndex     = { kind = 'index'  }
local Call            = { kind = 'call'   }

local Name            = { kind = 'name'   }
local Number          = { kind = 'number' }
local String          = { kind = 'string' }
local Bool            = { kind = 'bool'   }
  -- e.g. {0, 4, 3} {true, true, false}
local VectorLiteral   = { kind = 'vecliteral' } 

-- non-syntactic expressions (i.e. these will not be generated by the parser)
local FieldAccess     = { kind = 'fieldaccess' } -- type determined by field type
local Scalar          = { kind = 'scalar'      } -- type determined by scalar type

local QuoteExpr       = { kind = 'quoteexpr'   } -- type already checked, just return checked AST
local LuaObject       = { kind = 'luaobject' } --reference to a special Lua object, type is already provided


-- Statements:
local Statement       = { kind = 'statement'  }  -- abstract
  -- if expr then block (elseif cond then block)* (else block)? end
local IfStatement     = { kind = 'ifstmt'     }
local WhileStatement  = { kind = 'whilestmt'  }  -- while expr do block end
local DoStatement     = { kind = 'dostmt'     }  -- do block end
local RepeatStatement = { kind = 'repeatstmt' }  -- repeat block until cond
local ExprStatement   = { kind = 'exprstmt'   }  -- e;
local Assignment      = { kind = 'assnstmt'   }  -- "lvalue   = expr" 
local DeclStatement   = { kind = 'declstmt'   }  -- "var name"
local NumericFor      = { kind = 'numericfor' }
local GenericFor      = { kind = 'genericfor' }
local Break           = { kind = 'break'      }


----------------------------
--[[ Set up inheritance ]]--
----------------------------
local function inherit (child, parent)
	child.__index = child -- readies child as metatable for inheritance
	setmetatable(child, parent)
end

inherit(LisztKernel, AST)
inherit(Expression,  AST)
inherit(Statement,   AST)
inherit(Block,       AST)
inherit(CondBlock,   AST)

inherit(BinaryOp,      Expression)
inherit(UnaryOp,       Expression)
inherit(Number,        Expression)
inherit(String,        Expression)
inherit(Bool,          Expression)
inherit(Tuple,         Expression)
inherit(VectorLiteral, Expression)

inherit(Call,          Expression)
inherit(TableLookup,   Expression)
inherit(VectorIndex,   Expression)
inherit(Name,          Expression)
inherit(Reduce,        Expression)

inherit(Scalar,        Expression)
inherit(FieldAccess,   Expression)
inherit(QuoteExpr,     Expression)
inherit(LuaObject,     Expression)

inherit(IfStatement,     Statement)
inherit(WhileStatement,  Statement)
inherit(DoStatement,     Statement)
inherit(RepeatStatement, Statement)
inherit(ExprStatement,   Statement)
inherit(Assignment,      Statement)
inherit(DeclStatement,   Statement)
inherit(NumericFor,      Statement)
inherit(GenericFor,      Statement)
inherit(Break,           Statement)


-----------------------------
--[[ Lvalue flags        ]]--
-----------------------------
FieldAccess.is_lvalue = true
Scalar.is_lvalue      = true
Name.is_lvalue        = true

-----------------------------
--[[ General AST Methods ]]--
-----------------------------
function AST:New (P)
	local newnode = 
	{ 
		kind       = self.kind, 
		linenumber = P:cur().linenumber,
		filename   = P.source,
		offset     = P:cur().offset,
	}
	return setmetatable(newnode, self)
end

function AST:copy_location (node)
	self.linenumber = node.linenumber
	self.filename   = node.filename
	self.offset     = node.offset
end

function AST:DeriveFrom (ast)
	local newnode = setmetatable({kind=self.kind}, self)
	newnode:copy_location(ast)
	newnode.name      = ast.name
	newnode.node_type = self.node_type
	return newnode
end

function AST:clone ()
	local copy =
	{
		kind       = self.kind,
		linenumber = self.linenumber,
		filename   = self.filename,
		offset     = self.offset,
		is_lvalue  = self.is_lvalue,
		node_type  = self.node_type,
		name       = self.name,
	}
	return setmetatable(copy, getmetatable(self))
end

function Tuple:size ( ) return #self.children end

function AST:is (obj)
	return obj == getmetatable(self)
end

---------------------------
--[[ AST tree printing ]]--
---------------------------
local indent_delta = '   '

function LisztKernel:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (iter, set, body)")
	indent = indent .. indent_delta
	self.iter:pretty_print(indent)
	self.set:pretty_print(indent)
	self.body:pretty_print(indent)
end

function Block:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind)
	for i = 1, #self.statements do
		self.statements[i]:pretty_print(indent .. indent_delta)
	end
end

function CondBlock:pretty_print (indent)
	print(indent .. self.kind .. ": (cond, block)")
	self.cond:pretty_print(indent .. indent_delta)
	self.body:pretty_print(indent .. indent_delta)
end

function BinaryOp:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": " .. self.op)
	self.lhs:pretty_print(indent .. indent_delta)
	self.rhs:pretty_print(indent .. indent_delta)
end

function UnaryOp:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": " .. self.op)
	self.exp:pretty_print(indent .. indent_delta)
end

function Reduce:pretty_print(indent)
	indent = indent or ''
	print(indent .. self.kind .. ": " .. self.op)
	self.exp:pretty_print(indent .. indent_delta)
end

function Scalar:pretty_print(indent)
	indent = indent or ''
	local name = self.name or ""
	print(indent .. self.kind .. ": " .. name)
end

function QuoteExpr:pretty_print(indent)
    indent = indent or ''
    print(indent .. self.kind .. ": " .. tostring(self.ast))
end

function FieldAccess:pretty_print(indent)
	indent = indent or ''
	print(indent .. self.kind .. ': ' .. tostring(self.field))
end

function Tuple:pretty_print(indent)
	indent = indent or ''
	print(indent .. self.kind)
	for i = 1, #self.children do
		self.children[i]:pretty_print(indent .. indent_delta)
	end
end

function VectorLiteral:pretty_print(indent)
	indent = indent or ''
	print(indent .. self.kind)
	for i = 1, #self.elems do
		self.elems[i]:pretty_print(indent .. indent_delta)
	end
end

function Call:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (func, params)")
	self.func:pretty_print(indent .. indent_delta)
	self.params:pretty_print(indent .. indent_delta)
end

function TableLookup:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (table, member)")
	self.table:pretty_print(indent .. indent_delta)
	self.member:pretty_print(indent .. indent_delta)
end

function VectorIndex:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (vector, index)")
	self.vector:pretty_print(indent .. indent_delta)
	self.index:pretty_print(indent .. indent_delta)
end

function Name:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": " .. self.name)
end

function Number:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": " .. self.value)
end

function Bool:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ':' .. self.value)
end

function String:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": \"" .. self.value .. "\"")
end

function IfStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind)
	for i = 1, #self.if_blocks do
		self.if_blocks[i]:pretty_print(indent .. indent_delta)
	end
	if self.else_block then
		self.else_block:pretty_print(indent .. indent_delta)
	end
end

function WhileStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (condition, body)")
	self.cond:pretty_print(indent .. indent_delta)
	self.body:pretty_print(indent .. indent_delta)
end

function DoStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind)
	self.body:pretty_print(indent .. indent_delta)
end

function RepeatStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (body, condition)")
	self.body:pretty_print(indent .. indent_delta)
	self.cond:pretty_print(indent .. indent_delta)
end

function DeclStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind.. ":(typeexpression,initializer)")
	print(indent .. indent_delta .. self.name)
  if self.typeexpression then
      print(indent .. indent_delta ..self.typeexpression)
  end
  if self.initializer then
      self.initializer:print_pretty(indent .. indent_delta)
  end
end

function ExprStatement:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind)
	self.exp:pretty_print(indent .. indent_delta)
end

function Assignment:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ': (lvalue, exp)')
	self.lvalue:pretty_print(indent .. indent_delta)
	self.exp:pretty_print(indent .. indent_delta)
end

function Break:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind)
end

function NumericFor:pretty_print (indent)	
	indent = indent or ''
	if self.step then
		print(indent .. self.kind .. ": (iter, lower, upper, step, body)")
	else
		print(indent .. self.kind .. ": (iter, lower, upper, body)")
	end
	self.iter:pretty_print(indent .. indent_delta)
	self.lower:pretty_print(indent .. indent_delta)
	self.upper:pretty_print(indent .. indent_delta)
	if self.step then self.step:pretty_print(indent .. indent_delta) end
	self.body:pretty_print(indent .. indent_delta)
end

function GenericFor:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ": (iter, set, body)")
	self.iter:pretty_print(indent .. indent_delta)
	self.set:pretty_print(indent  .. indent_delta)
	self.body:pretty_print(indent .. indent_delta)
end

function VectorLiteral:pretty_print (indent)
	indent = indent or ''
	print(indent .. self.kind .. ":")
	for i = 1, #self.elems do
		self.elems[i]:pretty_print(indent .. indent_delta)
	end
end

-- declare other exports
for k,v in pairs({
	AST             = AST,
	LisztKernel     = LisztKernel,
	Block           = Block,
	Expression      = Expression,
	BinaryOp        = BinaryOp,
	UnaryOp         = UnaryOp,
	Reduce 	        = Reduce,
	Tuple           = Tuple,
	TableLookup     = TableLookup,
	VectorIndex     = VectorIndex,
	Scalar 	        = Scalar,
	FieldAccess     = FieldAccess,
	QuoteExpr       = QuoteExpr,
	Call            = Call,
	Name            = Name,
	Number          = Number,
	String          = String,
	Bool            = Bool,
	VectorLiteral   = VectorLiteral,
	Statement       = Statement,
	IfStatement     = IfStatement,
	WhileStatement  = WhileStatement,
	DoStatement     = DoStatement,
	RepeatStatement = RepeatStatement,
	ExprStatement   = ExprStatement,
	Assignment      = Assignment,
	DeclStatement   = DeclStatement,
	NumericFor      = NumericFor,
	GenericFor      = GenericFor,
	Break           = Break,
	CondBlock       = CondBlock,
	LuaObject       = LuaObject
}) do A[k] = v end

