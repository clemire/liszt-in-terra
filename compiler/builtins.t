local B = {}
package.loaded["compiler.builtins"] = B

local L = terralib.require "compiler.lisztlib"
local T = terralib.require "compiler.types"
local C = terralib.require "compiler.c"


---------------------------------------------
--[[ Builtin functions                   ]]--
---------------------------------------------
local Func = {}
Func.__index = Func
Func.__call = function(self,...)
    return self.luafunc(...)
end

function Func.new(luafunc)
    local check = function(ast, ctxt)
        error('unimplemented builtin function typechecking')
    end
    local codegen = function(ast, env)
        error('unimplemented builtin function codegen')
    end
    return setmetatable({check=check, codegen=codegen, luafunc = luafunc},
                        Func)
end
function B.isFunc(f)
    return getmetatable(f) == Func
end


B.assert = Func.new(assert)

function B.assert.check(ast, ctxt)
    local args = ast.params
    if #args ~= 1 then
        ctxt:error(ast, "assert expects exactly 1 argument (instead got " .. #args .. ")")
        return
    end

    local test = args[1]
    local test_type = test.node_type
    if test_type:isVector() then test_type = test_type:baseType() end
    if test_type ~= L.error and test_type ~= L.bool then
        ctxt:error(ast, "expected a boolean or vector of booleans as the test for assert statement")
    end
end

local terra lisztAssert(test : bool, file : rawstring, line : int)
    if not test then
        C.fprintf(C.get_stderr(), "%s:%d: assertion failed!\n", file, line)
        C.exit(1)
    end
end

function B.assert.codegen(ast, env)
    local test = ast.params[1]
    local code = test:codegen(env)
    if test.node_type:isVector() then
        if test.node_type.N == 0 then return quote end end
        local tt = test.node_type:terraType()
        local v = symbol()
        local alltest = `v[0]
        for i = 1, test.node_type.N - 1 do
            alltest = `alltest and v[i]
        end
        return quote var [v] = code in lisztAssert(alltest, ast.filename, ast.linenumber) end
    else
        return quote lisztAssert(code, ast.filename, ast.linenumber) end
    end
end


B.print = Func.new(print)

function B.print.check(ast, ctxt)
    local args = ast.params
    
    for i,output in ipairs(args) do
        local outtype = output.node_type
        if outtype ~= L.error and
           not outtype:isValueType() and not outtype:isRow()
        then
            ctxt:error(ast, "only numbers, bools, vectors and rows can be printed")
        end
    end
end

local function printOne(env,output)
    local lt   = output.node_type
    local tt   = lt:terraType()
    local code = output:codegen(env)
    if     lt == L.float or lt == L.double then return quote C.printf("%f", [double](code)) end
	elseif lt == L.int   or lt:isRow() then return quote C.printf("%d", code) end
	elseif lt == L.bool  then
        return quote C.printf("%s", terralib.select(code, "true", "false")) end
	elseif lt:isVector() then
        local printSpec = "{"
        local sym = symbol()
        local elemQuotes = {}
        local bt = lt:baseType()
        for i = 0, lt.N - 1 do
            if bt == L.float or bt == L.double then
                printSpec = printSpec .. " %f"
                table.insert(elemQuotes, `[double](sym[i]))
            elseif bt == L.int then
                printSpec = printSpec .. " %d"
                table.insert(elemQuotes, `sym[i])
            elseif bt == L.bool then
                printSpec = printSpec .. " %s"
                table.insert(elemQuotes, `terralib.select(sym[i], "true", "false"))
            else
                error('Unrecognized type in print: ' .. tostring(bt:terraType()))
            end
        end
        printSpec = printSpec .. " }"
        return quote
            var [sym] : tt = code
        in
            C.printf(printSpec, elemQuotes)
        end
    else
        assert(false and "printed object should always be number, bool, or vector")
    end
end
function B.print.codegen(ast, env)
    local output = ast.params[1]
    local stmts = {}
    for i,output in ipairs(ast.params) do
        table.insert(stmts, printOne(env,output))
        local t = ast.params[i+1] and " " or "\n"
        table.insert(stmts,`C.printf(t))
    end
	return stmts
end


local function dot(a, b)
    if not a.type:isVector() then
        error("first argument to dot must be a vector", 2)
    end
    if not b.type:isVector() then
        error("second argument to dot must be a vector", 2)
    end
    if #a.data ~= #b.data then
        error("cannot dot vectors of differing lengths (" ..
                #a.data .. " and " .. #b.data .. ")")
    end
    local sum = 0
    for i = 1, #a.data do
        sum = sum + a.data[i] * b.data[i]
    end
    return sum
end

B.dot = Func.new(dot)

function B.dot.check(ast, ctxt)
    local args = ast.params
    if #args ~= 2 then
        ctxt:error(ast, "dot expects exactly 2 arguments (instead got " .. #args .. ")")
        return
    end
    local lt1 = args[1].node_type
    local lt2 = args[2].node_type
    if not lt1:isVector() then
        ctxt:error(args[1], "first argument to dot must be a vector")
    end
    if not lt2:isVector() then
        ctxt:error(args[2], "second argument to dot must be a vector")
    end
    if not lt1:isVector() or not lt2:isVector() then return end
    if lt1.N ~= lt2.N then
        ctxt:error(ast, "cannot dot vectors of differing lengths (" ..
                lt1.N .. " and " .. lt2.N .. ")")
    end
    if lt1:baseType() == L.bool then
        ctxt:error(args[1], "boolean vector passed as first argument to dot")
    end
    if lt2:baseType() == L.bool then 
        ctxt:error(args[2], "boolean vector passed as second argument to dot")
    end
    return T.type_meet(lt1:baseType(), lt2:baseType())
end

function B.dot.codegen(ast, env)
    local args = ast.params
    local v1 = symbol()
    local v2 = symbol()
    local N = args[1].node_type.N
    if N == 0 then
        return `0
    end
    local result = `v1[0] * v2[0]
    -- TODO: make this codegen a Terra for loop for super-long vectors
    for i = 1, N - 1 do
        result = `result + v1[i] * v2[i]
    end

    local tt1 = args[1].node_type:terraType()
    local tt2 = args[2].node_type:terraType()
    local code1 = args[1]:codegen(env)
    local code2 = args[2]:codegen(env)
    return quote
        var [v1] : tt1 = code1
        var [v2] : tt2 = code2
    in
        result
    end
end


local function cross(a, b)
    if not a.type:isVector() then
        error("first argument to cross must be a vector", 2)
    end
    if not b.type:isVector() then
        error("second argument to cross must be a vector", 2)
    end
    local av = a.data
    if #av ~= 3 then
        error("arguments to cross must be vectors of length 3 (first argument is length " ..
                #av .. ")")
    end
    local bv = b.data
    if #bv ~= 3 then
        error("arguments to cross must be vectors of length 3 (second argument is length " ..
                #bv .. ")")
    end
    return L.NewVector(T.type_meet(a.type:baseType(), b.type:baseType()), {
        av[2] * bv[3] - av[3] * bv[2],
        av[3] * bv[1] - av[1] * bv[3],
        av[1] * bv[2] - av[2] * bv[1]
    })
end 

B.cross = Func.new(cross)

function B.cross.check(ast, ctxt)
    local args = ast.params
    if #args ~= 2 then
        ctxt:error(ast, "cross expects exactly 2 arguments (instead got " .. #args .. ")")
        return
    end
    local lt1 = args[1].node_type
    local lt2 = args[2].node_type
    if not lt1:isVector() then
        ctxt:error(args[1], "first argument to cross must be a vector")
    end
    if not lt2:isVector() then
        ctxt:error(args[2], "second argument to cross must be a vector")
    end
    if not lt1:isVector() or not lt2:isVector() then return end
    if lt1.N ~= 3 then
        ctxt:error(ast, "arguments to cross must be vectors of length 3 (first argument is of length" ..  lt1.N .. ")")
    end
    if lt2.N ~= 3 then
        ctxt:error(ast, "arguments to cross must be vectors of length 3 (second argument is of length" ..  lt1.N .. ")")
    end
    if lt1:baseType() == L.bool then
        ctxt:error(args[1], "boolean vector passed as first argument to cross")
    end
    if lt2:baseType() == L.bool then 
        ctxt:error(args[2], "boolean vector passed as second argument to cross")
    end
    return T.type_meet(lt1, lt2)
end

function B.cross.codegen(ast, env)
    local args = ast.params

    local tt1 = args[1].node_type:terraType()
    local tt2 = args[2].node_type:terraType()
    local v1 = symbol(tt1)
    local v2 = symbol(tt2)

    local tp = T.type_meet(args[1].node_type:baseType(), args[2].node_type:baseType()):terraType()
    local code1 = args[1]:codegen(env)
    local code2 = args[2]:codegen(env)
    return quote
        var [v1] = code1
        var [v2] = code2
    in
        vectorof(tp, v1[1] * v2[2] - v1[2] * v2[1],
                     v1[2] * v2[0] - v1[0] * v2[2],
                     v1[0] * v2[1] - v1[1] * v2[0])
    end
end


local function length(v)
    if not v.type:isVector() then
        error("argument to length must be a vector", 2)
    end
    return C.sqrt(dot(v, v))
end

B.length = Func.new(length)

function B.length.check(ast, ctxt)
    local args = ast.params
    if #args ~= 1 then
        ctxt:error(ast, "length expects exactly 1 argument (instead got " .. #args .. ")")
        return L.error
    end
    local lt = args[1].node_type
    if not lt:isVector() then
        ctxt:error(args[1], "argument to length must be a vector")
        return L.error
    end
    if lt:baseType() == L.bool then
        ctxt:error(args[1], "boolean vector passed as argument to length")
    end
    return L.float
end

function B.length.codegen(ast, env)
    local args = ast.params
    local sq = symbol()
    local N = args[1].node_type.N
    if N == 0 then
        return `0
    end
    local len2 = `sq[0]
    -- TODO: make this codegen a Terra for loop for super-long vectors
    for i = 1, N - 1 do
        len2 = `len2 + sq[i]
    end

    local tt = args[1].node_type:terraType()
    local code = args[1]:codegen(env)
    return quote
        var v : tt = code
        var [sq] : tt = v * v
    in
        C.sqrt(len2)
    end
end


local function map(fn, list)
    local result = {}
    for i = 1, #list do
        table.insert(result, fn(list[i]))
    end
    return result
end

local function GetTypedSymbol(arg)
    return symbol(arg.node_type:terraType())
end

local function TerraCheck(func)
    func:compile()
    return function (ast, ctxt)
        local args = ast.params
        local argsyms = map(GetTypedSymbol, args)
        local rettype = nil
        local try = function()
            local terrafunc = terra([argsyms]) return func([argsyms]) end
            terrafunc:compile()
            rettype = terrafunc:getdefinitions()[1]:gettype().returns
        end
        local success, retval = pcall(try)
        if not success then
            ctxt:error(ast, "couldn't fit parameters to signature of terra function")
            ctxt:error(ast, retval)
            return L.error
        end
        if #rettype > 1 then
            ctxt:error(ast, "terra function returns more than one value")
            return L.error
        end
        if #rettype == 1 and not T.terraToLisztType(rettype[1]) then
            ctxt:error(ast, "unable to use return type '" .. tostring(rettype[1]) .. "' of terra function in Liszt")
            return L.error
        end
        return #rettype == 0 and L.error or T.terraToLisztType(rettype[1])
    end
end

local function TerraCodegen(func)
    return function (ast, env)
        local args = ast.params
        local argsyms = map(GetTypedSymbol, args)
        local init_params = quote end
        for i = 1, #args do
            local code = args[i]:codegen(env)
            init_params = quote
                init_params
                var [argsyms[i]] = code
            end
        end
        return quote init_params in func(argsyms) end
    end 
end


function B.terra_to_func(terrafn)
    local newfunc = Func.new()
    newfunc.check, newfunc.codegen = TerraCheck(terrafn), TerraCodegen(terrafn)
    return newfunc
end

L.print  = B.print
L.assert = B.assert
L.dot    = B.dot
L.cross  = B.cross
L.length = B.length
L.is_function = B.isFunc
