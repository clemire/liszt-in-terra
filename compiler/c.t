-- location to put all C includes
-- the rest of the compiler should just terralib.require this file

-- Super ultra hack to get the directory of this file,
-- which can be mangled into the root Liszt directory.
-- Then we can make sure to stick the Liszt root on the
-- Clang path

-- Super Hacky!
local info = debug.getinfo(1, "S")
local src  = info.source
-- strip leading '@'' and trailing 'compiler/c.t'
local liszt_dir = src:sub(2,-14)

return terralib.includecstring([[
#include "runtime/src/lmeshloader.h"
#include "runtime/src/swiftshim.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>

// exposing pre-processor macro constants
int SEEK_SET_value () { return SEEK_SET; }
int SEEK_CUR_value () { return SEEK_CUR; }

FILE *get_stderr () { return stderr; }
]],
"-I"..liszt_dir
)
