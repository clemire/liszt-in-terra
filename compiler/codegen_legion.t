local Codegen = {}
package.loaded["compiler.codegen_legion"] = Codegen

local ast = require "compiler.ast"

local C = require 'compiler.c'
local L = require 'compiler.lisztlib'
local Cc = require 'compiler.codegen_common'

-- Legion dependencies
require "compiler.legion_data"
require "legionlib-terra"
local Lc = terralib.includecstring([[
#include "legion_c.h"
]])


-------------------------------------------------------------------------------
--[[                         Codegen Entrypoint                            ]]--
-------------------------------------------------------------------------------

function Codegen.codegen (kernel_ast, bran)
  local env  = terralib.newenvironment(nil)
  local ctxt = Context.new(env, bran)

  if ctxt:onGPU() then
    error("Unimplemented GPU codegen with Legion runtime")
  else
    return cpu_codegen_legion(kernel_ast, ctxt)
  end
end


-------------------------------------------------------------------------------
--[[                   Context Object for Compiler Pass                    ]]--
-------------------------------------------------------------------------------

local Context = Cc.Context

function Context:IndexSpaceIterator()
  -- TODO: 
  -- > Return a symbol for index space iterator corresponding to logical reion
  --   for relation over which kernel is invoked.
end

function Context:FieldAccessor(relation, field)
  -- TODO:
  -- > Return a symbol for accessing field from physical region for given pair
  --   of relation, field.
end


-------------------------------------------------------------------------------
--[[                         Codegen Entrypoint                            ]]--
-------------------------------------------------------------------------------

function Codegen.codegen (kernel_ast, bran)
  local env  = terralib.newenvironment(nil)
  local ctxt = Context.new(env, bran)

  if ctxt:onGPU() then
    error("Unimplemented GPU codegen with Legion runtime")
  else
    return cpu_codegen(kernel_ast, ctxt)
  end

end


----------------------------------------------------------------------------
--[[                         CPU Codegen                                ]]--
----------------------------------------------------------------------------

function cpu_codegen (kernel_ast, ctxt)
  ctxt:enterblock()
    -- declare the symbol for iteration

    -- body of kernel loop
    local body = quote
      -- [kernel_ast.body:codegen(ctxt)]
    end

    -- Placeholder for kernel executable.
    local kernel_body = quote
      C.printf("Kernel executable unimplemented\n")
      return false
    end

    -- special iteration logic for subset-mapped kernels
    if ctxt.bran.subset then
      error("Unimplemented subsets codegen with Legion runtime")
    else
      -- kernel_body = quote
      --   TODO:
      --   > Define index space iterator using logical region for the relation
      --   > Iterate over index space
      --     [body]
      --   end
    end
  ctxt:leaveblock()

  local k = terra (regions : &Lc.legion_physical_region_t,
                   num_regions : uint32,
                   ctx : Lc.legion_context_t,
                   runtime : Lc.legion_runtime_t)
    [kernel_body]
  end
  k:setname(kernel_ast.id)
  return k
end
