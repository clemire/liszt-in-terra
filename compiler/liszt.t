--this file is used to extend the terra parser with
--the liszt language
--use it like: import "compiler.liszt"

local P = require "compiler.parser"

local lisztlib = terralib.require "compiler.lisztlib"
local semant = terralib.require "compiler.semant"
-- include liszt library for programmer
L = lisztlib

local lisztlanguage = {
	name        = "liszt", -- name for debugging
	entrypoints = {"liszt_kernel", "liszt"},
	keywords    = {"var", "kernel"},

	expression = function(self, lexer)
		local ast = P.Parse(lexer)
        if ast.kind == 'kernel' then
            return function (env_fn) 
                local env = env_fn()
                return lisztlib.NewKernel(ast, env)
            end
        else
            return function (env_fn)
                local env = env_fn()
                return semant.check(env, ast)
            end
        end
	end
}

return lisztlanguage
