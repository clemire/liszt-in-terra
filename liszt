#!/bin/bash

# one liner from stack overflow that gets the
# full resident directory path of this script
SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
# The above allows us to run this script from most anywhere as a program
# If something breaks, uncomment the diagnostic echo and check
# Stack Overflow for more ideas.  There were also elaborate multi-line
# solutions to handle various edge-cases
#echo $SCRIPT_DIR


OPTS="-p$SCRIPT_DIR/?.t -p$SCRIPT_DIR/?.lua"
LISZT_RUNTIME=$SCRIPT_DIR/runtime/libsingle_runtime.so
#echo $OPTS

# the following line is complaining about a unary operator, what gives?
if [ $1 == "-v" ];
then
	LISZT_RUNTIME=$LISZT_RUNTIME $SCRIPT_DIR/terra/terra -vv $OPTS $2
else
	LISZT_RUNTIME=$LISZT_RUNTIME $SCRIPT_DIR/terra/terra  $OPTS $1
fi
