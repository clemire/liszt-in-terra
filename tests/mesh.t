import "compiler.liszt"

local LMesh = terralib.require "compiler.lmesh"
local M = LMesh.Load("examples/mesh.lmesh")

M.vertices:print()
M.faces:print()
M.edges:print()
M.cells:print()
M.edgesofface:print()
M.edgesofcell:print()
M.verticesofface:print()