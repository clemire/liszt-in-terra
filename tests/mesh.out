vertices	size: 8
position: <vector(double,3)>
	0	1 1 1 
	1	1 1 -1 
	2	1 -1 1 
	3	1 -1 -1 
	4	-1 1 1 
	5	-1 1 -1 
	6	-1 -1 1 
	7	-1 -1 -1 
faces	size: 6
outside: <uint64>
	0	0
	1	0
	2	0
	3	0
	4	0
	5	0
inside: <uint64>
	0	1
	1	1
	2	1
	3	1
	4	1
	5	1
edges	size: 12
head: <uint64>
	0	2
	1	3
	2	1
	3	0
	4	6
	5	7
	6	5
	7	4
	8	4
	9	1
	10	6
	11	3
tail: <uint64>
	0	0
	1	2
	2	3
	3	1
	4	4
	5	6
	6	7
	7	5
	8	0
	9	5
	10	2
	11	7
cells	size: 2
edgesofface	size: 24
face: <uint64>
	0	0
	1	0
	2	0
	3	0
	4	1
	5	1
	6	1
	7	1
	8	2
	9	2
	10	2
	11	2
	12	3
	13	3
	14	3
	15	3
	16	4
	17	4
	18	4
	19	4
	20	5
	21	5
	22	5
	23	5
edge: <uint64>
	0	0
	1	1
	2	2
	3	3
	4	4
	5	5
	6	6
	7	7
	8	8
	9	7
	10	9
	11	3
	12	10
	13	5
	14	11
	15	1
	16	8
	17	4
	18	10
	19	0
	20	9
	21	6
	22	11
	23	2
orientation: <bool>
	0	false
	1	false
	2	false
	3	false
	4	false
	5	false
	6	false
	7	false
	8	false
	9	true
	10	false
	11	false
	12	false
	13	false
	14	false
	15	true
	16	false
	17	false
	18	true
	19	true
	20	true
	21	true
	22	false
	23	false
edgesofcell	size: 24
cell: <uint64>
	0	0
	1	0
	2	0
	3	0
	4	0
	5	0
	6	1
	7	1
	8	1
	9	1
	10	1
	11	1
	12	1
	13	1
	14	1
	15	1
	16	1
	17	1
	18	1
	19	1
	20	1
	21	1
	22	1
	23	1
edge: <uint64>
	0	0
	1	1
	2	6
	3	7
	4	9
	5	10
	6	0
	7	1
	8	2
	9	2
	10	3
	11	3
	12	4
	13	4
	14	5
	15	5
	16	6
	17	7
	18	8
	19	8
	20	9
	21	10
	22	11
	23	11
verticesofface	size: 24
face: <uint64>
	0	0
	1	0
	2	0
	3	0
	4	1
	5	1
	6	1
	7	1
	8	2
	9	2
	10	2
	11	2
	12	3
	13	3
	14	3
	15	3
	16	4
	17	4
	18	4
	19	4
	20	5
	21	5
	22	5
	23	5
vertex: <uint64>
	0	0
	1	2
	2	3
	3	1
	4	4
	5	6
	6	7
	7	5
	8	0
	9	4
	10	5
	11	1
	12	2
	13	6
	14	7
	15	3
	16	0
	17	4
	18	6
	19	2
	20	1
	21	5
	22	7
	23	3
